### ISSUE TITLE
UrlMatcher as a service

### ISSUE THREAD

**Issue by PleasantD** [angular/angular#17145]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Only a function can be set as the `urlMatcher` of a route.

```
const routes = [{
    urlMatcher: (segments: UrlSegment[], segmentGroup: UrlSegmentGroup, route: Route) => {
        /* Do stuff */
    },
    component: MyComponent,
}]
```

**Expected behavior**
A DI-resolved service or function can be set as the `urlMatcher` allowing for further injection. I'd expect this to work in an identical way to how the resolver properties can either be services that implement the `Resolve<T>` interface OR can be an inline `resolve()` function.

```
const routes = [{
    urlMatcher: MyUrlMatcherService,
    component: InterruptComponent,
},
{
    path: ":id",
    component: ListComponent, /* Contains a <router-outlet> for routing children */ 
    children: [ /*children with sub-paths for deep linking */ ]
}];

@Injectable()
export class MyUrlMatcherService {
    constructor(myStateService:MyStateService){}

    matchUrl(segments: UrlSegment[], segmentGroup: UrlSegmentGroup, route: Route):UrlMatchResult {
        if (!this.myStateService.isInterrupt) return null;
        /* Do matching stuff */
    }
}
```

**Minimal reproduction of the problem with instructions**
N/A

**What is the motivation / use case for changing the behavior?**
My use case is that `InterruptComponent` should resolve if some system state (managed by a service) is currently true. If not, I want to skip this route an let other routes attempt matching. In my case I want to greedily consume the entire sub-path and display a specific component, but only in certain circumstances.

**Please tell us about your environment:** 
N/A

* **Angular version:** 4.X
* **Browser:** N/A
* **Language:** TypeScript 2.2+ 
* **Node (for AoT issues):** N/A

**Comment by Dok11** [comment]
Hi all! Have any news about this brilliant feature?

**Comment by MatissJanis** [comment]
Would be absolutely amazing if we could have this!

**Comment by korenb** [comment]
Also, it would be nice to have an ability to return `Promise<UrlMatchResult >` in `matchUrl` function

**Comment by korenb** [comment]
@Dok11 , you're right! it should support Observable. Because sometimes i need to call api before matcher has resolved

**Comment by adrigardi90** [comment]
I agree! If the matchUrl function could support asynchronous behaviour, would be really good

**Comment by EddyP23** [comment]
@korenb, do you have a workaround for the case when you need to call API before the matcher has resolved?

**Comment by EddyP23** [comment]
and +1 for the feature, would be very useful

**Comment by AmandaFoster** [comment]
@EddyP23 
In our application, we need information about a user from a remote source to determine which route to use. We load the user at login time and cache it in a service, as well as storing a global reference to the Angular injector. The matcher imports the injector, finds the user service, and returns the cached user through `Observable.of()`, which is synchronous. It's a bit hacky, but it works.

In short, the route which needs an asynchronous matcher would need to be preceded by a route that doesn't, and that initial route would need to get the remote information via a resolver or a service injected into a component.

If your workflow doesn't give you the opportunity to preload the data, you could have a route that presents a loading screen with a distinct URL, do the async request, and redirect to the route with the async matcher.

**Comment by joselcvarela** [comment]
It's been a year since last comment and I would love to have this feature on my application.
Specifically I would want to inject Ngrx Store into matcher so I can conditionally match the route accordingly to my store.

**Comment by gilsdav** [comment]
Add Injector as parameter of UrlMatcher seems to be quickly implementable.
It will allow us to use services into a matcher.
Is it a viable solution ?

**Comment by SachinShekhar** [comment]
I also need this feature badly. It'd be much easier to inject other services into this than grabbing Injector instance from main.ts. Plus, asynchronous path matching is a necessity. Unfortunately, matcher doesn't take async functions.

Example Use Case: A URL segment can represent a tweet, user, group and I want to lazily load a module only after finding out what this segment is post API call.

**Comment by william-lohan** [comment]
Why not `CanActivate`?

**Comment by amandabot** [comment]
Our particular use case is that we load a different module and component for the same route depending on the user's type.

`CanActivate` and `Resolve` are triggered after the route is matched, so you can't use them if the route is conditional on asynchronous data.

**Comment by lazarljubenovic** [comment]
I thought I could do a lazy match of slugs based on a service where I fetch observables from... but apparently not, because 

- I cannot return an observable from the matcher and
- I cannot even inject a service.

In a framework driven by observables and dependency injection, finding something that suddenly doesn't have either is a real downer. To find an ancient, well-upovted issue without a single official comment by a team members is an even bigger downer.

It's been two and a half years...

**Comment by xavadu** [comment]
That would be perfect, so matcher can be used just as guards or resolvers.

Additionally it looks that matcher and resolver can be at the same time, not specified in the documentation, the resolver is ignored. 

Retrieve information or 'resolve' is a responsibility of the resolver, instead that the matcher return a paramMap would be good if it is used as a :parameter which at the end, a matcher is used when a :parameter is not enough, then matcher -> guard/resolver -> component.

**Comment by krokofant** [comment]
Struggling with this for quite a while, I wonder if this can be partially solved by the `Router.resetConfig` method?

Injecting the router service in the router module and then conditionally add the route configurations... 🤔

**Comment by lukasz-madej** [comment]
Would love to see that feature implemented. This would solve all current issues I have with implementing a 404 page for missing products / articles. Having a 404 page rendered without changing the url an keeping the "back" button functionality i key for me.

**Comment by atscott** [comment]
It seems like all of the use-cases here would be covered by #14515. Is there a use-case here that's wouldn't be covered by that?

**Comment by atscott** [comment]
@krokofant Could you elaborate on why #14515 is not a suitable replacement for this request. I understand that these two are concretely different in the implementation details, but the question is whether there are good examples of cases where dynamically changing the result of `UrlMatchResult` is desired (outside of a dynamic `null` vs the static `UrlMatchResult`). Why not create separate `UrlMatcher`s that do one thing rather than return several different `UrlMatchResult`s? This seems better than a `UrlMatcher` that composes the result in different ways based on various conditions.

**Comment by OysteinAmundsen** [comment]
I have an example where I need this:

In my application I use a rather complex tree structured hierarchy of nodes of different types. The currently selected node is available from a service and each time a node is selected its entire path is rendered in the url as query params. The reason for this, is that the selected node structure is used accross several different routes. But the queryparams is not necessarily ordered as this can be invoked externally as well, so I depend on having the service available for determining which node is the currently selected one. For this one route in my app, which is for configuring nodes of specific types, each node type has its specific component.

Right now, I do this is done inside a component wrapper where I run `const node = service.getSelectedNode()` and in the template several `*ngIf="node.type === 'someType'` to invoke the destination component.

What I would like to do is:

```
export const configRoutes: Routes = [
  {
    path: '',
    component: ConfigurationComponent,
    // FIXME: https://github.com/angular/angular/issues/17145
    children: [
      { matcher: createNodeMatcher('root'), component: RootFormComponent},
      { matcher: createNodeMatcher('assetGroup'), component: AssetGroupComponent},
      { matcher: createNodeMatcher('asset'), component: AssetFormComponent},
      { matcher: createNodeMatcher('section'), component: SectionFormComponent},
      { matcher: createNodeMatcher('level'), component: LevelFormComponent},
      { matcher: createNodeMatcher('zone'), component: ZoneFormComponent},
      { matcher: createNodeMatcher('room'), component: RoomFormComponent},
      { matcher: createNodeMatcher('resource'), component: ResourceFormComponent},
      { matcher: createNodeMatcher('assetMeteringPointGroup'), component: MeterGroupComponent },
    ],
  },
];

export function createNodeMatcher(match: string): UrlMatcher {
  return (segments: UrlSegment[], group: UrlSegmentGroup, route: Route): UrlMatchResult | null => {
    // Need to inject the `service` somehow
    return service.getSelectedNode().type === match ? { consumed: segments } : null;
  };
}

export const ConfigurationRoutes = RouterModule.forChild(configRoutes);
```
But the `AssetTreeService` needs to be injected somehow.

Now that I think about it, can I configure the RouterModule injected to the module from the module constructor (where the injector is available)?
